linkerd/distribute/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+//! A load-agnostic traffic distribution stack module.
+
+#![deny(rust_2018_idioms, clippy::disallowed_methods, clippy::disallowed_types)]
+#![forbid(unsafe_code)]
+
+use indexmap::IndexMap;
+use linkerd_stack::{NewService, Param, Service};
+use rand::{
+    distributions::{Distribution as _, WeightedIndex},
+    rngs::SmallRng,
+    SeedableRng,
+};
+use std::{
+    collections::HashSet,
+    hash::Hash,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+#[derive(Clone, Debug)]
+pub struct NewDistribute<K, S> {
+    backends: Arc<IndexMap<K, S>>,
+}
+
+#[derive(Debug)]
+pub struct Distribute<K, S> {
+    backends: Arc<IndexMap<K, S>>,
+    ready: Option<S>,
+    inner: Inner<K>,
+}
+
+/// A parameter type that configures how a [`Distribute`] should behave.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Distribution<K> {
+    /// A distribution that has no backends, and therefore never becomes ready.
+    Empty,
+
+    /// A distribution that uses the first available backend in an ordered list.
+    FirstAvailable(Arc<[K]>),
+
+    /// A distribution that uses the first available backend when randomly
+    /// selecting over a weighted distributino of backends.
+    RandomAvailable(Arc<WeightedKeys<K>>),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct WeightedKeys<K> {
+    keys: Vec<K>,
+    index: WeightedIndex<u32>,
+}
+
+#[derive(Debug)]
+enum Inner<K> {
```

Comment:
naming nit, take it or leave it: if we wanted to give this a more descriptive name, what do you think of `Selection`?